vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/text_processing/embedding_processing.cc

```diff
@@ -0,0 +1,131 @@
+#include "bat/ads/internal/ml/pipeline/pipeline_embedding_info.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_util.h"
+#include "bat/ads/internal/ml/pipeline/text_processing/embedding_data.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace ads {
+namespace ml {
+namespace pipeline {
+
+// static
+std::unique_ptr<EmbeddingProcessing> EmbeddingProcessing::CreateFromValue(
+    base::Value resource_value,
+    std::string* error_message) {
+  DCHECK(error_message);
+
+  auto text_processing = std::make_unique<EmbeddingProcessing>();
+  if (!text_processing->FromValue(std::move(resource_value))) {
+    *error_message = "Failed to parse text embedding pipeline JSON";
+    return {};
+  }
+
+  return text_processing;
+}
+
+bool EmbeddingProcessing::IsInitialized() const {
+  return is_initialized_;
+}
+
+EmbeddingProcessing::EmbeddingProcessing() : is_initialized_(false) {}
+
+EmbeddingProcessing::~EmbeddingProcessing() = default;
+
+void EmbeddingProcessing::SetInfo(const PipelineEmbeddingInfo& info) {
+  embedding_pipeline_.version = info.version;
+  embedding_pipeline_.timestamp = info.timestamp;
+  embedding_pipeline_.locale = info.locale;
+  embedding_pipeline_.embeddings_dim = info.embeddings_dim;
+  embedding_pipeline_.embeddings = info.embeddings;
+}
+
+bool EmbeddingProcessing::FromValue(base::Value resource_value) {
+  absl::optional<PipelineEmbeddingInfo> pipeline_embedding_info =
+      ParsePipelineEmbedding(std::move(resource_value));
+
+  if (pipeline_embedding_info.has_value()) {
+    SetInfo(pipeline_embedding_info.value());
+    is_initialized_ = true;
+  } else {
+    is_initialized_ = false;
+  }
+
+  return is_initialized_;
+}
+
+std::string EmbeddingProcessing::CleanText(const std::string& text,
+                                           bool is_html) {
+  std::string cleaned_text = text;
+  if (is_html) {
+    cleaned_text = ParseTagAttribute(cleaned_text, "og:title", "content");
+  }
+  cleaned_text = StripNonAlphaCharacters(cleaned_text);
+  std::transform(cleaned_text.begin(), cleaned_text.end(), cleaned_text.begin(),
+                 ::tolower);
+  return cleaned_text;
+}
+
+TextEmbeddingData EmbeddingProcessing::EmbedText(
+    const std::string& text) const {
+  std::vector<float> embedding_initialize(embedding_pipeline_.embeddings_dim,
+                                          0.0);
+  VectorData embedding_vector = VectorData(embedding_initialize);
+  TextEmbeddingData embedding_data;
+  embedding_data.embedding = embedding_vector;
+
+  if (!IsInitialized()) {
+    return embedding_data;
+  }
+
+  const std::vector<std::string> tokens = base::SplitString(
+      text, " ", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+
+  std::string in_vocab_text;
+  float n_tokens = 0.0;
+  for (const auto& token : tokens) {
+    const auto iter = embedding_pipeline_.embeddings.find(token);
+    if (iter != embedding_pipeline_.embeddings.end()) {
+      BLOG(1, token << " - token found");
```

Comment:
nit: could we remove these verbose BLOGs before merging?

Reply:
I think those might be useful for debugging/QAing this. Maybe we could increase verbosity to e.g. 3?

Reply:
Maybe 7, as this way they do not appear in rewards diagnostic logs? Or would that be helpful too

Reply:
For now I went with 3 for BLOGs related to the new incoming embedding; 7 for BLOG-ing the stored history

Reply:
@ptjames can you please paste an example log for level 3. Thanks

Reply:
For level 3 logging on `embedding_processing.cc` and `text_embedding_processor.cc` (only new incoming embedding detailed):

`[63202:259:0727/112120.757677:VERBOSE3:embedding_processing.cc(118)] nowhere
[63202:259:0727/112120.757747:VERBOSE3:embedding_processing.cc(118)] to
[63202:259:0727/112120.757771:VERBOSE3:embedding_processing.cc(113)] live - token found
[63202:259:0727/112120.757793:VERBOSE3:embedding_processing.cc(118)] rents
[63202:259:0727/112120.757813:VERBOSE3:embedding_processing.cc(118)] in
[63202:259:0727/112120.757834:VERBOSE3:embedding_processing.cc(118)] canada
[63202:259:0727/112120.757853:VERBOSE3:embedding_processing.cc(118)] surge
[63202:259:0727/112120.757872:VERBOSE3:embedding_processing.cc(118)] as
[63202:259:0727/112120.757891:VERBOSE3:embedding_processing.cc(113)] home - token found
[63202:259:0727/112120.757910:VERBOSE3:embedding_processing.cc(118)] prices
[63202:259:0727/112120.757929:VERBOSE3:embedding_processing.cc(118)] fall
[63202:259:0727/112120.757953:VERBOSE3:text_embedding_processor.cc(69)] Embedding: 0.012099800631403923 0.003527998924255371 0.04225249961018562 0.010532599873840809 0.0702265053987503 -0.03969299793243408 0.02259249985218048 -0.14045999944210052 0.026814499869942665 0.012325899675488472 -0.07007399946451187 -0.11089999973773956 -0.013682451099157333 -0.0033355001360177994 0.008434300310909748 -0.008263849653303623 0.173895001411438 -0.009649000130593777 0.09451200067996979 0.03579520061612129 0.02942650020122528 -0.040190499275922775 0.020244833081960678 0.07758349925279617 0.03385449945926666 0.02438800036907196 -0.002101849764585495 0.051614001393318176 -0.02787790074944496 0.04722999781370163 -0.05781400203704834 0.014968350529670715 -0.05002199858427048 -0.014060551300644875 0.03179050236940384 0.0171050988137722 -0.0006434500683099031 -0.00900100078433752 0.02047634869813919 0.019228998571634293 -0.00760085042566061 -0.11702549457550049 0.0015054999385029078 -0.04063500091433525 -0.02675740048289299 -0.03420349955558777 -0.0012838500551879406 -0.03716849908232689 -0.02599100023508072 0.05341250076889992 -0.00665650051087141 0.03276599943637848 0.016215499490499496 -0.007932500913739204 -0.18819500505924225 0.0066673001274466515 0.013117000460624695 -0.012672499753534794 -0.07459750026464462 -0.06158750131726265 -0.022819150239229202 0.0032950006425380707 0.20921500027179718 0.047177501022815704 0.0675479993224144 0.01692900061607361 -0.05902250111103058 -0.029422501102089882 -0.013617800548672676 -0.0551145002245903 0.04506799951195717 0.06539499759674072 0.06785400211811066 -0.03357619792222977 0.015912599861621857 0.016298498958349228 0.031938500702381134 -0.029961001127958298 -0.02061450108885765 0.0340379998087883 0.017947498708963394 -0.03937999904155731 0.03429799899458885 0.07367250323295593 -0.020963501185178757 -0.04421699792146683 -0.017939649522304535 -0.07105249911546707 -0.009775649756193161 0.002049999311566353 -0.04834970086812973 0.052293501794338226 -0.15592500567436218 0.06680049747228622 0.00938700046390295 0.050857000052928925 0.04844849929213524 0.054978497326374054 0.06337550282478333 -0.020068200305104256 0.0918550044298172 -0.005845045205205679 0.02495879866182804 -0.03260450065135956 0.06364399939775467 -0.16652999818325043 -0.029615500941872597 0.021351700648665428 0.041735999286174774 -0.02910515107214451 -0.047613997012376785 0.07776334881782532 0.09650100022554398 0.07043799757957458 -0.01657950133085251 0.027577999979257584 -0.02334675006568432 -0.024020899087190628 -0.04479549825191498 -0.02400900050997734 -0.060412999242544174 0.030789850279688835 0.002617998979985714 0.008192850276827812 0.01186390034854412 -0.006045499816536903 0.006483498960733414 -0.1077210009098053 0.04778800159692764 0.03695850074291229 0.0715159997344017 0.05210749804973602 0.08642999827861786 -0.04616899788379669 -0.014154849573969841 -0.023644499480724335 0.04335799813270569 0.07131099700927734 -0.018419450148940086 -0.02098745107650757 0.024692248553037643 -0.014053000137209892 0.03587950021028519 0.01771249994635582 -0.03904715180397034 0.06368549913167953 0.0004450008273124695 0.0458344966173172 -0.021736999973654747 -0.057157501578330994 -0.007065599784255028 0.12171250581741333 -0.036979999393224716 -0.0701964944601059 -0.019609497860074043 -0.037053998559713364 0.03379850089550018 0.10523200035095215 0.07780300080776215 -0.07112349569797516 0.03187499940395355 0.015434999018907547 0.024876149371266365 0.04594700038433075 0.015141800977289677 -0.0027599986642599106 0.021990999579429626 0.007691498845815659 -0.02732899971306324 0.024268848821520805 -0.018911000341176987 0.030758000910282135 0.01562650129199028 0.03748200088739395 -0.0038845017552375793 0.0505170002579689 0.06945699453353882 -0.014258500188589096 -0.05781799927353859 0.010350500233471394 0.08053000271320343 0.012721499428153038 -0.016111500561237335 -0.026677999645471573 -0.059119999408721924 -0.03989744931459427 -0.04948100075125694 0.09876199811697006 -0.012593700550496578 -0.07803699374198914 0.0647135004401207 0.0676954984664917 -0.007663000375032425 0.000733410008251667 -0.03412500023841858 -0.007415000349283218 0.0663049966096878 0.048601001501083374 0.022344401106238365 0.029467999935150146 -0.23083999752998352 0.19308499991893768 -0.019517000764608383 -0.0825280025601387 -0.05185500159859657 -0.028532501310110092 -0.022969000041484833 -0.04023599997162819 -0.017690999433398247 -0.03511900082230568 0.09664499759674072 -0.08768299967050552 0.004576500039547682 0.0076538994908332825 -0.03896833956241608 -0.04655659943819046 -0.0006610006093978882 0.0320190005004406 0.06050560250878334 -0.07280249893665314 0.04357500001788139 -0.03338349983096123 0.004113500006496906 0.1732950061559677 0.008971000090241432 -0.042634446173906326 -0.02882070094347 0.11692649871110916 -0.08411699533462524 -0.053760502487421036 0.061477500945329666 -0.08076299726963043 -0.058579497039318085 -0.03217549994587898 -0.020730499178171158 0.005080000031739473 0.014157800003886223 0.05060650035738945 0.0005399999208748341 -0.013178499415516853 0.16934999823570251 -0.02444482035934925 0.03392600268125534 -0.004909999668598175 -0.12325900048017502 0.030910350382328033 -0.00639289990067482 -0.06295400112867355 -0.007615000009536743 0.010429499670863152 0.048606500029563904 -0.03266235068440437 -0.0315450020134449 -0.016942501068115234 -0.02699434943497181 -0.013733499683439732 -0.0066979993134737015 0.027592498809099197 -0.018740501254796982 -0.011712498962879181 0.06071379780769348 0.007226998917758465 -0.030878450721502304 -0.07600099593400955 0.05580650269985199 -0.008987000212073326 0.0964754968881607 -0.027321498841047287 0.052367500960826874 0.014817500486969948 -0.06252799928188324 0.1299000084400177 -0.1737300008535385 -0.07734650373458862 -0.01053400058299303 -0.05377649888396263 0.019968200474977493 0.0037945001386106014 0.05357249826192856 0.04436265304684639 0.07176899909973145 0.0033435001969337463 -0.024486999958753586 -0.01726396381855011 0.02203909493982792 0.008879500441253185 0.04826590046286583 -0.06842649728059769 -0.022105999290943146 -0.009549899958074093 0.047918498516082764 0.07180599868297577 -0.057482000440359116 0.003404499962925911 0.004313749726861715 0.010204450227320194 0.03901199996471405 0.03664175048470497 -0.027098000049591064 0.06154000014066696`

Reply:
This verbose logging should be added to level 9 so that it is not part of rewards diagnostic logs.

Reply:
Changed to level 9

Reply:
Hmmm I did leave a comment as to why. Rewards diagnostic logs contain verbose levels 0 to 6.